examples/Fine_tuning_for_function_calling.ipynb

```diff
@@ -0,0 +1,1248 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Getting baseline function calling performance"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# !pip install tenacity\n",
+    "# !pip insta openai\n",
+    "# !pip install typing\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 33,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import openai\n",
+    "import numpy as np\n",
+    "import json\n",
+    "import os\n",
+    "import openai\n",
+    "import itertools\n",
+    "import random\n",
+    "from tenacity import retry, wait_random_exponential, stop_after_attempt,wait_fixed\n",
+    "from typing import Iterable, Any, Union, Dict, List, Generator\n",
+    "import ast\n",
+    "GPT_MODEL = 'gpt-4'\n",
+    "openai.api_key = os.getenv('OPENAI_API_KEY')\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Utilities"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's define utility functions for making calls to the Chat Completions API, one to get the completion and one to get the function call."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@retry(wait=wait_random_exponential(min=1, max=20), stop=stop_after_attempt(5))\n",
+    "def get_function_call(\n",
+    "    messages: list[dict[str, str]],\n",
+    "    model: str = \"gpt-4\",\n",
+    "    max_tokens=500,\n",
+    "    temperature=1.0,\n",
+    "    stop=None,\n",
+    "    functions=None,\n",
+    ") -> str:\n",
+    "    completion = openai.ChatCompletion.create(\n",
+    "        model=model,\n",
+    "        messages=messages,\n",
+    "        max_tokens=max_tokens,\n",
+    "        temperature=temperature,\n",
+    "        stop=stop,\n",
+    "        functions=functions,\n",
+    "    )\n",
+    "    return completion.choices[0].message\n",
+    "\n",
+    "@retry(wait=wait_random_exponential(min=1, max=20), stop=stop_after_attempt(5))\n",
+    "def get_chat_completion(\n",
+    "    messages: list[dict[str, str]],\n",
+    "    model: str = \"gpt-4\",\n",
+    "    max_tokens=500,\n",
+    "    temperature=1.0,\n",
+    "    stop=None,\n",
+    ") -> str:\n",
+    "    completion = openai.ChatCompletion.create(\n",
+    "        model=model,\n",
+    "        messages=messages,\n",
+    "        max_tokens=max_tokens,\n",
+    "        temperature=temperature,\n",
+    "        stop=stop,\n",
+    "    )\n",
+    "    return completion.choices[0].message\n",
+    "\n"
```

Comment:
Any reason not to just have a single one used both for functions and completions?

Reply:
For whatever reason using using the function calling with functions=None was failing, may have just been a me issue